File: MercuryAPI-1.31.3.36/java/samples_nb/src/samples/RegionConfiguration.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package samples;
import com.thingmagic.*;
import java.util.Arrays;

/**
 *
 * @author sghatak
 */
public class RegionConfiguration {
    
    /**Open region parameters
     * Set custom values by amending default values below.
     */
    static boolean lbtEnable = true;
    static boolean dwellTimeEnable = true;
    /**
     * dwellTime value Min 1 - Max 65525
     * 0 is not valid for dwellTime. 
     **/
    static int dwellTime = 100;
    /**
     * lbtThreshold value should be between -128 to 0
     * LBT threshold default value is -72.
     **/
    static int lbtThreshold = -72;
    /**
     * hopTime, quantizationStep, minFrequency & hopTable to be changed according to region.
     **/
    static int hopTime = 3675;
    static int quantizationStep = 100000;
    static int minFrequency = 865700;
    static int[] hopTable = {865700, 866300, 866900, 867500};
    
    public enum ConfigRegion
    {
        DEFAULT(0),
        BAHRAIN(1),
        BRAZIL(2);
        int value;
        ConfigRegion(int v)
        {
          value = v;
        }
        public int getValue()
        {
          return value;
        }
        
    };
    
    /**
     * Enum region codes:
     * 0 - Default
     * 1 - Bahrain
     * 2 - Brazil
     */
    static ConfigRegion confRegion = ConfigRegion.DEFAULT;
    static void usage()
    {
      System.out.printf("Usage: Please provide valid arguments, such as:\n"
                  + "read [-v] [reader-uri] [--ant n[,n...]] [--lbtThreshold] [LBT threshold value] [--dwellTime] [dwell time value] \n" +
                    "-v  Verbose: Turn on transport listener\n" +
                    "reader-uri  Reader URI: e.g., \"tmr:///COM1\", \"tmr://astra-2100d3\"\n"
                  + "--ant  Antenna List: e.g., \"--ant 1\", \"--ant 1,2\"\n"
                  + "--region region value: e.g., \"--region Brazil\" or \"--region Bahrain\" or do not pass this argument for default values\n"
                  + "e.g: tmr:///com1 --ant 1,2 --lbtThreshold -72 --dwellTime 100 --region Brazil\n ");
      System.exit(1);
    }

    static void setRegionParams(ConfigRegion region)
    {
        switch(region)
        {
            //Amend case 0 for custome values
            case DEFAULT:
                //default values - set to Bharain region values
                System.out.println("********Setting default region params********"+"\n");
                break;
            case BAHRAIN:
                //default values will hold for Bharain region
                System.out.println("********Setting params for Bahrain region********"+"\n");
                lbtEnable = true;
                lbtThreshold = -72;
                dwellTimeEnable = true;
                dwellTime = 100;
                hopTime = 3675;
                quantizationStep = 100000;
                minFrequency = 865700;
                //hop table for Bahrain region
                hopTable = new int[]{865700, 866300, 866900, 867500};
                break;
            case BRAZIL:
                System.out.println("********Setting params for Brazil region********"+"\n");
                lbtEnable = false;
                lbtThreshold = -72;
                dwellTimeEnable = false;
                dwellTime = 1000;
                hopTime = 375;
                quantizationStep = 250000;
                minFrequency = 902000;
                //hop table for Brazil region
                hopTable = new int[]{918250,923250,927750,905250,923750,918750,926250,921250,905750,915250,904750,902250,
                916750,926750,921750,925250,916250,922750,904250,917250,903750,906250,919750,927250,922250,920750,925750,
                920250,924750,915750,903250,919250,924250,902750,917750,906750};
                break;
        }
    }
    
    public static void setTrace(Reader r, String args[])
    {    
      if (args[0].toLowerCase().equals("on"))
      {
        r.addTransportListener(r.simpleTransportListener);
      }    
    }
 
    public static void main(String argv[])
    {
      // Program setup
      Reader r = null;
      int nextarg = 0;
      boolean trace = false;
      int[] antennaList = null;

      if (argv.length < 1)
        usage();

      if (argv[nextarg].equals("-v"))
      {
        trace = true;
        nextarg++;
      }

      // Create Reader object, connecting to physical device
      try
      {

          TagReadData[] tagReads;
          String readerURI = argv[nextarg];
          nextarg++;

          for (; nextarg < argv.length; nextarg++)
          {
              String arg = argv[nextarg];
              if (arg.equalsIgnoreCase("--ant"))
              {
                  if (antennaList != null)
                  {
                      System.out.println("Duplicate argument: --ant specified more than once");
                      usage();
                  }
                  antennaList = parseAntennaList(argv, nextarg);
                  nextarg++;
              }
              else if (arg.equalsIgnoreCase("--region"))
              {
                  confRegion = parseArguments(argv, nextarg);
                  if(confRegion == ConfigRegion.DEFAULT)
                  {
                      System.out.println("Argument " + argv[nextarg+1] + " is not recognised. "+"Supported regions : \"--region Bahrain\" & \"--region Brazil\"");
                      usage();
                  }
                  nextarg++;
              }
              else
              {
                  System.out.println("Argument " + argv[nextarg] + " is not recognised");
                  usage();
              }

          }

          r = Reader.create(readerURI);
          if (trace)
          {
              setTrace(r, new String[] {"on"});
          }
          r.connect();
          if (Reader.Region.UNSPEC == (Reader.Region) r.paramGet("/reader/region/id"))
          {
              Reader.Region[] supportedRegions = (Reader.Region[]) r.paramGet(TMConstants.TMR_PARAM_REGION_SUPPORTEDREGIONS);
              if (supportedRegions.length < 1)
              {
                  throw new Exception("Reader doesn't support any regions");
              }
              else
              {
                  r.paramSet("/reader/region/id", supportedRegions[0]);
              }
          }
          //Step 1 - Set region to OPEN
          r.paramSet(TMConstants.TMR_PARAM_REGION_ID, Reader.Region.OPEN);
          Reader.Region region = (Reader.Region) r.paramGet(TMConstants.TMR_PARAM_REGION_ID);
          if(region == Reader.Region.OPEN)
          {
              System.out.println("\n"+"********Region set to OPEN********"+"\n");
              //Set region parameters as per the selected region
              setRegionParams(confRegion);
              //Step 2 - Set frequency hop table
              System.out.println("********Setting Hop table********");
              r.paramSet(TMConstants.TMR_PARAM_REGION_HOPTABLE, hopTable);
              int[] hopTableVal = (int[])r.paramGet(TMConstants.TMR_PARAM_REGION_HOPTABLE);
              System.out.println("Hop table set to : " +Arrays.toString(hopTableVal)+"\n");
              //Step 3 - Set LBT enable
              System.out.println("********Setting LBT enable********");
              r.paramSet(TMConstants.TMR_PARAM_REGION_LBT_ENABLE, lbtEnable);
              //Get & print LBT enable value
              boolean lbtEnableVal = (Boolean) r.paramGet(TMConstants.TMR_PARAM_REGION_LBT_ENABLE);
              System.out.println("LBT enable : "+lbtEnableVal+"\n");
              //Step 4 - Set LBT threshold
              if(lbtEnable)
              {
                System.out.println("********Setting LBT threshold********");
                r.paramSet(TMConstants.TMR_PARAM_REGION_LBT_THRESHOLD, lbtThreshold);
                //Get & print LBT threshold value
                int lbtThresholdVal = (Integer) r.paramGet(TMConstants.TMR_PARAM_REGION_LBT_THRESHOLD);
                System.out.println("LBT threshold set to : "+lbtThresholdVal+" dBm"+"\n");
              }
              else
              {
                  System.out.println("LBT enable set to false, not setting LBT threshold"+"\n");
              }
              //Step 5 - Set dwell time enable
              System.out.println("********Setting dwell time enable********");
              r.paramSet(TMConstants.TMR_PARAM_REGION_DWELL_TIME_ENABLE, dwellTimeEnable);
              //Get & print dwell time enable value
              boolean dwellTimeEnableVal = (Boolean) r.paramGet(TMConstants.TMR_PARAM_REGION_DWELL_TIME_ENABLE);
              System.out.println("Dwell time enable : "+dwellTimeEnableVal+"\n");
              //Step 6 - Set dwell time
              if(dwellTimeEnable)
              {
                System.out.println("********Setting dwell time********");
                r.paramSet(TMConstants.TMR_PARAM_REGION_DWELL_TIME, dwellTime);//Get & print dwell time value
                int dwellTimeVal = (Integer) r.paramGet(TMConstants.TMR_PARAM_REGION_DWELL_TIME);
                System.out.println("Dwell time set to : "+dwellTimeVal+" ms"+"\n");
              }
              else
              {
                  System.out.println("Dwell time enable set to false, not setting dwell time"+"\n");
              }
              //Step 7 - Set hop time
              System.out.println("********Setting Hop time********");
              r.paramSet(TMConstants.TMR_PARAM_REGION_HOPTIME, hopTime);
              int hopTimeVal = (Integer) r.paramGet(TMConstants.TMR_PARAM_REGION_HOPTIME);
              System.out.println("Hop time set to : "+hopTimeVal+" ms"+"\n");
              //Step 8 - Set quantization step
              System.out.println("********Setting Quantization step********");
              r.paramSet(TMConstants.TMR_PARAM_REGION_QUANTIZATION_STEP, quantizationStep);
              int quantizationStepVal = (Integer) r.paramGet(TMConstants.TMR_PARAM_REGION_QUANTIZATION_STEP);
              System.out.println("Quantization step set to - "+((quantizationStepVal/1000))+" kHz"+"\n");
              //Step 9 - Set region minimum frequency
              System.out.println("********Setting minimum frequency********");
              r.paramSet(TMConstants.TMR_PARAM_REGION_MINIMUM_FREQUENCY, minFrequency);
              int minFrequencyVal = (Integer) r.paramGet(TMConstants.TMR_PARAM_REGION_MINIMUM_FREQUENCY);
              System.out.println("Minimum region frequency set to : "+minFrequencyVal+"\n");

              /**
               * Uncomment Step 10 through 12 to save OPEN region settings as persistent.
               */
              //Step 10 - Save configuration
              //r.paramSet("/reader/userConfig", new SerialReader.UserConfigOp(SerialReader.SetUserProfileOption.SAVE));
              //System.out.println("User profile set option:save all configuration");
              //Step 11 - Restore configuration
              //r.paramSet("/reader/userConfig", new SerialReader.UserConfigOp(SerialReader.SetUserProfileOption.RESTORE));
              //System.out.println("User profile set option:restore all configuration");
              //Step 12 - Verify configuration
              //r.paramSet("/reader/userConfig", new SerialReader.UserConfigOp(SerialReader.SetUserProfileOption.VERIFY));
              //System.out.println("User profile set option:verify all configuration");
          }
          else
          {
              System.out.println("Region configuration is only supported for OPEN region");
              r.destroy();
              System.exit(1);
          }

          if (r.isAntDetectEnabled(antennaList))
          {
              System.out.println("Module doesn't has antenna detection support, please provide antenna list");
              r.destroy();
              usage();
          }
          SimpleReadPlan plan = new SimpleReadPlan(antennaList, TagProtocol.GEN2, null, null, 1000);
          r.paramSet("/reader/read/plan", plan);
          // Read tags
          tagReads = r.read(1000);
          // Print tag reads
          for (TagReadData tr : tagReads)
          {
              System.out.println("EPC: " + tr.epcString());
          }
          // Shut down reader
          r.destroy();
      } 
      catch (ReaderException re)
      {
        System.out.println("Reader Exception: " + re.getMessage());
      }
      catch (Exception re)
      {
          System.out.println("Exception: " + re.getMessage());
      }
    }
  
    static int[] parseAntennaList(String[] args,int argPosition)
    {
        int[] antennaList = null;
        try
        {
            String argument = args[argPosition + 1];
            String[] antennas = argument.split(",");
            int i = 0;
            antennaList = new int[antennas.length];
            for (String ant : antennas)
            {
                antennaList[i] = Integer.parseInt(ant);
                i++;
            }
        }
        catch (IndexOutOfBoundsException ex)
        {
            System.out.println("Missing argument after " + args[argPosition]);
            usage();
        }
        catch (Exception ex)
        {
            System.out.println("Invalid argument at position " + (argPosition + 1) + ". " + ex.getMessage());
            usage();
        }
        return antennaList;
    }
  
    static ConfigRegion parseArguments(String[] args,int argPosition)
    {
        ConfigRegion reg = ConfigRegion.DEFAULT;
        try
        {
            String param = args[argPosition];
            String argument = args[argPosition + 1];
            if(param.equalsIgnoreCase("--region"))
            {
                String region = argument.toUpperCase();
                if(region.equalsIgnoreCase("Bahrain"))
                {
                    reg = ConfigRegion.BAHRAIN;
                }
                else if(region.equalsIgnoreCase("Brazil"))
                {
                    reg = ConfigRegion.BRAZIL;
                }
                else
                {
                    reg = ConfigRegion.DEFAULT;
                }
            }
            else
            {
                reg = ConfigRegion.DEFAULT;
            }
        }
        catch (IndexOutOfBoundsException ex)
        {
            System.out.println("Missing argument after " + args[argPosition]);
            usage();
        }
        catch (Exception ex)
        {
            System.out.println("Invalid argument at position " + (argPosition + 1) + ". " + ex.getMessage());
            usage();
        }
        return reg;
    }
}

File: MercuryAPI-1.31.3.36/c/src/samples/RegionConfiguration.c
/**
 * Sample program that demonstrates open region configuration 
 * @file RegionConfiguration.c
 */

#include <tm_reader.h>
#include <time.h>
#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <string.h>
#include <inttypes.h>
#include <serial_reader_imp.h>

#ifndef BARE_METAL
#if WIN32
#define snprintf sprintf_s
#endif 

/* Enable this to use transportListener */
#ifndef USE_TRANSPORT_LISTENER
#define USE_TRANSPORT_LISTENER 0
#endif
#define PRINT_TAG_METADATA 0
#define numberof(x) (sizeof((x))/sizeof((x)[0]))

#define usage() {errx(1, "Please provide valid arguments, such as: reader-uri [--ant n] [--region BRAZIL/Brazil/brazil]\n"\
                         "reader-uri : e.g., 'tmr:///COM1' or 'tmr:///dev/ttyS0/' or 'tmr://readerIP'\n"\
                         "[--ant n] : e.g., '--ant 1'\n"\
                         "[--region region] : e.g, '--region Brazil'\n"\
                         "Example: 'tmr:///com4 --ant 1,2 --region Brazil'\n");}

void printU32List(TMR_uint32List *list)
{
  int i;
  putchar('[');
  for (i = 0; i < list->len && i < list->max; i++)
  {
    printf("%"PRIu32"%s", list->list[i], ((i + 1) == list->len) ? "" : ",");
  }
  if (list->len > list->max)
  {
    printf("...");
  }
  putchar(']');
}

void errx(int exitval, const char *fmt, ...)
{
  va_list ap;

  va_start(ap, fmt);
  vfprintf(stderr, fmt, ap);

  exit(exitval);
}

void checkerr(TMR_Reader* rp, TMR_Status ret, int exitval, const char *msg)
{
  if (TMR_SUCCESS != ret)
  {
    errx(exitval, "Error %s: %s\n", msg, TMR_strerr(rp, ret));
  }
}

void serialPrinter(bool tx, uint32_t dataLen, const uint8_t data[],
                   uint32_t timeout, void *cookie)
{
  FILE *out = cookie;
  uint32_t i;

  fprintf(out, "%s", tx ? "Sending: " : "Received:");
  for (i = 0; i < dataLen; i++)
  {
    if (i > 0 && (i & 15) == 0)
    {
      fprintf(out, "\n         ");
    }
    fprintf(out, " %02x", data[i]);
  }
  fprintf(out, "\n");
}

void stringPrinter(bool tx,uint32_t dataLen, const uint8_t data[],uint32_t timeout, void *cookie)
{
  FILE *out = cookie;

  fprintf(out, "%s", tx ? "Sending: " : "Received:");
  fprintf(out, "%s\n", data);
}

void parseAntennaList(uint8_t *antenna, uint8_t *antennaCount, char *args)
{
  char *token = NULL;
  char *str = ",";
  uint8_t i = 0x00;
  int scans;

  /* get the first token */
  if (NULL == args)
  {
    fprintf(stdout, "Missing argument\n");
    usage();
  }

  token = strtok(args, str);
  if (NULL == token)
  {
    fprintf(stdout, "Missing argument after %s\n", args);
    usage();
  }

  while(NULL != token)
  {
    scans = sscanf(token, "%"SCNu8, &antenna[i]);
    if (1 != scans)
    {
      fprintf(stdout, "Can't parse '%s' as an 8-bit unsigned integer value\n", token);
      usage();
    }
    i++;
    token = strtok(NULL, str);
  }
  *antennaCount = i;
}
#endif

int main(int argc, char *argv[])
{
  TMR_Reader r, *rp;
  TMR_Status ret;
  TMR_ReadPlan plan;
  TMR_Region region;
  uint8_t *antennaList = NULL;
  uint8_t buffer[20];
  uint8_t i, setRegion = 0;
  uint8_t antennaCount = 0x0;
  TMR_RegionList regions;
  TMR_Region _regionStore[32];
  TMR_uint32List value;
  //hopTime, quantizationStep & minFrequency to be changed according to region
  uint32_t valueList[4] = { 865700,866300,866900,867500 };
  uint32_t hopTime = 100;
  uint32_t quantizationStep = 100000; /* 100 KHz */ 
  uint32_t minFrequency = 865700;
  bool lbtEnable  = true;
  bool dwellTimeEnable = true;
  /* dwellTime value Min 1 - Max 65535.
   * dwellTime default value is 100
   */
  uint32_t dwellTime = 100;
  /* LBTThreshold value should be between -128 to -1.
   * LBT default value is -72
   */
  int8_t lbtThreshold  = -72;
  enum configRegion
  {
    TMR_REG_BAHRAIN = 1,
    TMR_REG_BRAZIL = 2
  };

#ifndef BARE_METAL
#if USE_TRANSPORT_LISTENER
  TMR_TransportListenerBlock tb;
#endif
 
  if (argc < 2)
  {
    fprintf(stdout, "Not enough arguments.  Please provide reader URL.\n");
    usage(); 
  }

  for (i = 2; i < argc; i+=2)
  {
    if(0x00 == strcmp("--ant", argv[i]))
    {
      if (NULL != antennaList)
      {
        fprintf(stdout, "Duplicate argument: --ant specified more than once\n");
        usage();
      }
      parseAntennaList(buffer, &antennaCount, argv[i+1]);
      antennaList = buffer;
    }
    else if (0 == strcmp("--region", argv[i]))
    {
      if ((0 == strcmp("bahrain", argv[i+1])) || (0 == strcmp("Bahrain", argv[i+1])) || (0 == strcmp("BAHRAIN", argv[i+1])))
      {
        setRegion = 1;
      }
      else if ((0 == strcmp("brazil", argv[i+1])) || (0 == strcmp("Brazil", argv[i+1])) || (0 == strcmp("BRAZIL", argv[i+1])))
      {
        setRegion = 2;
      }
      else
      {
        fprintf(stdout, "Can't parse region value : %s\n,Please make manual changes in codelete to support custom region!", argv[i+1]);
      }
    }
    else
    {
      fprintf(stdout, "Argument %s is not recognized\n", argv[i]);
      usage();
    }
  }
#endif

  rp = &r;
#ifndef BARE_METAL
  ret = TMR_create(rp, argv[1]);
#else
  ret = TMR_create(rp, "tmr:///com1");
  buffer[0] = 1;
  antennaList = buffer;
  antennaCount = 0x01;
#endif

#ifndef BARE_METAL
  checkerr(rp, ret, 1, "creating reader");

#if USE_TRANSPORT_LISTENER

  if (TMR_READER_TYPE_SERIAL == rp->readerType)
  {
    tb.listener = serialPrinter;
  }
  else
  {
    tb.listener = stringPrinter;
  }
  tb.cookie = stdout;

  TMR_addTransportListener(rp, &tb);
#endif
#endif

  ret = TMR_connect(rp);

#ifndef BARE_METAL
  checkerr(rp, ret, 1, "connecting reader");
#endif
  region = TMR_REGION_NONE;
  ret = TMR_paramGet(rp, TMR_PARAM_REGION_ID, &region);
#ifndef BARE_METAL
  checkerr(rp, ret, 1, "getting region");
#endif

  if (TMR_REGION_NONE == region)
  {
    regions.list = _regionStore;
    regions.max = sizeof(_regionStore)/sizeof(_regionStore[0]);
    regions.len = 0;

    ret = TMR_paramGet(rp, TMR_PARAM_REGION_SUPPORTEDREGIONS, &regions);
#ifndef BARE_METAL
    checkerr(rp, ret, __LINE__, "getting supported regions");

    if (regions.len < 1)
    {
      checkerr(rp, TMR_ERROR_INVALID_REGION, __LINE__, "Reader doesn't support any regions");
    }
#endif
    region = regions.list[0];
    ret = TMR_paramSet(rp, TMR_PARAM_REGION_ID, &region);
#ifndef BARE_METAL
    checkerr(rp, ret, 1, "setting region");
#endif
  }

  /**
   * Checking the software version of the sargas.
   * The antenna detection is supported on sargas from software version of 5.3.x.x.
   * If the sargas software version is 5.1.x.x then antenna detection is not supported.
   * User has to pass the antenna as arguments.
   */
  {
    ret = isAntDetectEnabled(rp, antennaList);
    if(TMR_ERROR_UNSUPPORTED == ret)
    {
#ifndef BARE_METAL
      fprintf(stdout, "Reader doesn't support antenna detection. Please provide antenna list.\n");
      usage();
#endif
    }
    else
    {
#ifndef BARE_METAL
      checkerr(rp, ret, 1, "Getting Antenna Detection Flag Status");
#endif
    }
  }
  value.max = numberof(valueList);
  value.len = numberof(valueList);
  value.list = valueList;

  //Step 1 - Set region to OPEN
  region = TMR_REGION_OPEN;
  ret = TMR_paramSet(rp, TMR_PARAM_REGION_ID, &region);
  checkerr(rp, ret, 1, "setting region");

  /* Step 2 - Set frequency hop table accordingly for a region. */
  switch (setRegion)
  {
    case TMR_REG_BAHRAIN:
    {
      uint32_t hopTable[4] = { 865700,866300,866900,867500 };
      value.max = numberof(hopTable);
      value.len = numberof(hopTable);
      value.list = hopTable;

      lbtEnable  = true;
      dwellTimeEnable = true;
      hopTime = 100; /* 100ms */
      quantizationStep = 100000; /* 100 KHz */
      minFrequency = 865700; /* 902MHz */
      dwellTime = 100;
      lbtThreshold = -72;
    }
    break;
    case TMR_REG_BRAZIL:
    {
      uint32_t brazilHopTable[36] = {902250, 902750, 903250, 903750, 904250, 904750, 905250, 905750, 906250, 906750,
                                     915250, 915750, 916250, 916750, 917250, 917750, 918250, 918750, 919250, 919750, 
                                     920250, 920750, 921250, 921750, 922250, 922750, 923250, 923750, 924250, 924750,
                                     925250, 925750, 926250, 926750, 927250, 927750};
      value.max = numberof(brazilHopTable);
      value.len = numberof(brazilHopTable);
      value.list = brazilHopTable;

      hopTime = 375; /* 375ms */
      quantizationStep = 250000; /* 250 KHz */
      minFrequency = 902000; /* 902MHz */
      lbtEnable  = false;
      dwellTimeEnable = false;
    }
    break;
    default:
    {
      printf("\nSetting default region params...!\n");
    }
    break;
  }
  {
    ret = TMR_paramSet(rp, TMR_PARAM_REGION_HOPTABLE, &value);
    checkerr(rp, ret, 1, "Setting Hoptable");
    ret = TMR_paramGet(rp, TMR_PARAM_REGION_HOPTABLE, &value);
    checkerr(rp, ret, 1, "New Hoptable");
    putchar('\n');
    printU32List(&value);
    putchar('\n');
  }

  //Set lbtEnable value to true or false
  ret = TMR_paramSet(rp, TMR_PARAM_REGION_LBT_ENABLE, &lbtEnable);  //Step 3 - Set LBT enable
  checkerr(rp, ret, 1, "Setting LBT Enable");
  //Set lbtThreshold value only if lbtEnable is set to true.
  if (lbtEnable)
  {
    ret = TMR_paramSet(rp, TMR_PARAM_REGION_LBT_THRESHOLD, &lbtThreshold);  //Step 4 - Set LBT threshold
    checkerr(rp, ret, 1, "Setting LBT Threshold");
  }

  //Set dwellTimeEnable value to true or false
  ret = TMR_paramSet(rp, TMR_PARAM_REGION_DWELL_TIME_ENABLE, &dwellTimeEnable);  //Step 5 - Set Dwell time enable
  checkerr(rp, ret, 1, "Setting Dwell Time Enable");
  //Set dwellTime value only if dwellTimeEnable is set to true.
  if(dwellTimeEnable)
  {
    ret = TMR_paramSet(rp, TMR_PARAM_REGION_DWELL_TIME, &dwellTime);  //Step 6 -Set Dwell time
    checkerr(rp, ret, 1, "Setting Dwell Time");
  }

  //Get & print LBT enable
  ret = TMR_paramGet(rp, TMR_PARAM_REGION_LBT_ENABLE, &lbtEnable);
  checkerr(rp, ret, 1, "Getting LBT En");
  printf("LBT Enable: %d \n", lbtEnable);
  //Get & print LBT threshold
  ret = TMR_paramGet(rp, TMR_PARAM_REGION_LBT_THRESHOLD, &lbtThreshold);
  checkerr(rp, ret, 1, "Setting LBT Threshold");
  printf("LBT Threshold(dBm): %d \n", lbtThreshold);
  //Get & print Dwell time enable
  ret = TMR_paramGet(rp, TMR_PARAM_REGION_DWELL_TIME_ENABLE, &dwellTimeEnable);
  checkerr(rp, ret, 1, "Setting Dwell Time Enable");
  printf("Dwell Time Enable: %d \n", dwellTimeEnable);
  //Get & print Dwell time
  ret = TMR_paramGet(rp, TMR_PARAM_REGION_DWELL_TIME, &dwellTime);
  checkerr(rp, ret, 1, "Setting Dwell Time");
  printf("Dwell Time(ms): %d \n", dwellTime);

  //Step 7 - Set hop time
  ret = TMR_paramSet(rp, TMR_PARAM_REGION_HOPTIME, &hopTime);
  checkerr(rp, ret, 1, "Setting Hop Time");
  //Get & print hop time
  ret = TMR_paramGet(rp, TMR_PARAM_REGION_HOPTIME, &hopTime);
  checkerr(rp, ret, 1, "Getting Hop Time");
  printf("Hop Time: %d\n", hopTime);

  //Step 8 - Set Quantization Step
  TMR_paramSet(rp, TMR_PARAM_REGION_QUANTIZATION_STEP, &quantizationStep);
  checkerr(rp, ret, 1, "Setting Quantization Step");
  //Get & print Quantization Step
  TMR_paramGet(rp, TMR_PARAM_REGION_QUANTIZATION_STEP, &quantizationStep);
  checkerr(rp, ret, 1, "Setting Quantization Step");
  printf("Quantization Step: %d\n", quantizationStep);

  //Step 9 - Set Minimum Frequency
  TMR_paramSet(rp, TMR_PARAM_REGION_MINIMUM_FREQUENCY, &minFrequency);
  checkerr(rp, ret, 1, "Setting Minimum Frequency");
  //Get & print Minimum Frequency
  TMR_paramGet(rp, TMR_PARAM_REGION_MINIMUM_FREQUENCY, &minFrequency);
  checkerr(rp, ret, 1, "Setting Minimum Frequency");
  printf("Minimum Frequency: %d\n", minFrequency);

  /* Uncomment Step 10 to Step 12 to save OPEN region settings as persistent. */
  //{
  //  TMR_SR_UserConfigOp config;
  //  //Step 10 - Init UserConfigOp structure to save configuration
  //  TMR_init_UserConfigOp(&config, TMR_USERCONFIG_SAVE);
  //  ret = TMR_paramSet(rp, TMR_PARAM_USER_CONFIG, &config);
  //  checkerr(rp, ret, 1, "setting user configuration: save all configuration");
  //  printf("User config set option:save all configuration\n");

  //  //Step 11 - Init UserConfigOp structure to Restore all saved configuration parameters
  //  TMR_init_UserConfigOp(&config, TMR_USERCONFIG_RESTORE);
  //  ret = TMR_paramSet(rp, TMR_PARAM_USER_CONFIG, &config);
  //  checkerr(rp, ret, 1, "setting configuration: restore all saved configuration params");
  //  printf("User config set option:restore all saved configuration params\n");

  //  //Step 12 - Init UserConfigOp structure to verify all saved configuration parameters
  //  TMR_init_UserConfigOp(&config, TMR_USERCONFIG_VERIFY);
  //  ret = TMR_paramSet(rp, TMR_PARAM_USER_CONFIG, &config);
  //  checkerr(rp, ret, 1, "setting configuration: verify all saved configuration params");
  //  printf("User config set option:verify all configuration\n");
  //}

  //Initialize the read plan
  ret = TMR_RP_init_simple(&plan, antennaCount, antennaList, TMR_TAG_PROTOCOL_GEN2, 1000);
#ifndef BARE_METAL
  checkerr(rp, ret, 1, "initializing the  read plan");
#endif

  //Commit read plan
  ret = TMR_paramSet(rp, TMR_PARAM_READ_PLAN, &plan);
#ifndef BARE_METAL
  checkerr(rp, ret, 1, "setting read plan");
#endif
  ret = TMR_read(rp, 500, NULL);

#ifndef BARE_METAL
if (TMR_ERROR_TAG_ID_BUFFER_FULL == ret)
  {
    /* In case of TAG ID Buffer Full, extract the tags present
    * in buffer.
    */
    fprintf(stdout, "reading tags:%s\n", TMR_strerr(rp, ret));
  }
  else
  {
    checkerr(rp, ret, 1, "reading tags");
  }
#endif
  while (TMR_SUCCESS == TMR_hasMoreTags(rp))
  {
    TMR_TagReadData trd;
    char epcStr[128];

    ret = TMR_getNextTag(rp, &trd);
#ifndef BARE_METAL  
  checkerr(rp, ret, 1, "fetching tag");
#endif
    TMR_bytesToHex(trd.tag.epc, trd.tag.epcByteCount, epcStr);
    printf("EPC: %s\n ", epcStr);
  }

  TMR_destroy(rp);
  return 0;
}
